### ISSUE TITLE
BEGIN blocks have wrong caller package

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#15597]
Migrated from [rt.perl.org#129239](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=129239) (status was 'open')

Searchable as RT129239$

**Comment by p5pRT** [comment]
### From @mauke

#### Created by @mauke



\#\!perl
use strict;
use warnings;

package Mtfnpy;
\#line 100 "ABCDE"
BEGIN \{
&nbsp;    printf "package=%s\, file=%s\, line=%d\\n"\, caller;
\}

\_\_END\_\_

Output&#8203;:
package=main\, file=ABCDE\, line=102

Expected output&#8203;:
package=Mtfnpy\, file=ABCDE\, line=102

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.24.0:

Configured by mauke at Mon May  9 21:21:33 CEST 2016.

Summary of my perl5 (revision 5 version 24 subversion 0) configuration:
   
  Platform:
    osname=linux, osvers=4.4.5-1-arch, archname=i686-linux
    uname='linux simplicio 4.4.5-1-arch #1 smp preempt thu mar 10 07:54:30 cet 2016 i686 gnulinux '
    config_args=''
    hint=previous, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -flto',
    cppflags='-fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
    ccversion='', gccversion='6.1.1 20160501', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234, doublekind=3
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12, longdblkind=3
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags ='-fstack-protector-strong -L/usr/local/lib -flto'
    libpth=/usr/local/lib /usr/lib/gcc/i686-pc-linux-gnu/6.1.1/include-fixed /usr/lib /lib /usr/local/lib /usr/lib/gcc/i686-pc-linux-gnu/6.1.1/include-fixed /usr/lib
    libs=-lpthread -lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc -lgdbm_compat
    perllibs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    libc=libc-2.23.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.23'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -flto -L/usr/local/lib -fstack-protector-strong'



@INC for perl 5.24.0:
    /home/mauke/usr/lib/perl5/site_perl/5.24.0/i686-linux
    /home/mauke/usr/lib/perl5/site_perl/5.24.0
    /home/mauke/usr/lib/perl5/5.24.0/i686-linux
    /home/mauke/usr/lib/perl5/5.24.0
    .


Environment for perl 5.24.0:
    HOME=/home/mauke
    LANG=en_US.UTF-8
    LANGUAGE=en_US
    LC_COLLATE=C
    LC_MONETARY=de_DE.UTF-8
    LC_TIME=de_DE.UTF-8
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/mauke/perl5/perlbrew/bin:/home/mauke/bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl
    PERLBREW_BASHRC_VERSION=0.73
    PERLBREW_HOME=/home/mauke/.perlbrew
    PERLBREW_ROOT=/home/mauke/perl5/perlbrew
    PERL_BADLANG (unset)
    PERL_UNICODE=SAL
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Sat Sep 10 02&#8203;:13&#8203;:50 2016\, mauke\- wrote&#8203;:
> 
> This is a bug report for perl from l\.mai@&#8203;web\.de\,
> generated with the help of perlbug 1\.40 running under perl 5\.24\.0\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> \#\!perl
> use strict;
> use warnings;
> 
> package Mtfnpy;
> \#line 100 "ABCDE"
> BEGIN \{
>     printf "package=%s\, file=%s\, line=%d\\n"\, caller;
> \}
> 
> \_\_END\_\_
> 
> Output&#8203;:
> package=main\, file=ABCDE\, line=102
> 
> Expected output&#8203;:
> package=Mtfnpy\, file=ABCDE\, line=102

I’m not entirely sure I would expect that\.  I’m not sure what to expect\.  After all\, ‘calling’ BEGIN in this case has not started running yet\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @mauke

On Sat Sep 10 07&#8203;:31&#8203;:47 2016\, sprout wrote&#8203;:
> On Sat Sep 10 02&#8203;:13&#8203;:50 2016\, mauke\- wrote&#8203;:
> > \#\!perl
> > use strict;
> > use warnings;
> >
> > package Mtfnpy;
> > \#line 100 "ABCDE"
> > BEGIN \{
> >     printf "package=%s\, file=%s\, line=%d\\n"\, caller;
> > \}
> >
> > \_\_END\_\_
> >
> > Output&#8203;:
> > package=main\, file=ABCDE\, line=102
> >
> > Expected output&#8203;:
> > package=Mtfnpy\, file=ABCDE\, line=102
> 
> I’m not entirely sure I would expect that\.  I’m not sure what to
> expect\.  After all\, ‘calling’ BEGIN in this case has not started
> running yet\.

I don't understand your last sentence but the reported location of the call \(which is the "\}" of the BEGIN block\) is in package Mtfnpy\, not main\. At minimum it's inconsistent\.

**Comment by p5pRT** [comment]
### From @iabyn

On Sat\, Sep 10\, 2016 at 07&#8203;:47&#8203;:42AM \-0700\, l\.mai@&#8203;web\.de via RT wrote&#8203;:
> On Sat Sep 10 07&#8203;:31&#8203;:47 2016\, sprout wrote&#8203;:
> > On Sat Sep 10 02&#8203;:13&#8203;:50 2016\, mauke\- wrote&#8203;:
> > > \#\!perl
> > > use strict;
> > > use warnings;
> > >
> > > package Mtfnpy;
> > > \#line 100 "ABCDE"
> > > BEGIN \{
> > >     printf "package=%s\, file=%s\, line=%d\\n"\, caller;
> > > \}
> > >
> > > \_\_END\_\_
> > >
> > > Output&#8203;:
> > > package=main\, file=ABCDE\, line=102
> > >
> > > Expected output&#8203;:
> > > package=Mtfnpy\, file=ABCDE\, line=102
> > 
> > I’m not entirely sure I would expect that\.  I’m not sure what to
> > expect\.  After all\, ‘calling’ BEGIN in this case has not started
> > running yet\.
> 
> I don't understand your last sentence but the reported location of the call \(which is the "\}" of the BEGIN block\) is in package Mtfnpy\, not main\. At minimum it's inconsistent\.

What's happening is as follows&#8203;:

At compile time\, 'package Foo' sets PL\_curstash to point to %Foo&#8203;::\.
Then any subsequent COPs \(i\.e\. OP\_NEXTSTATE\) after that have their
CopSTASH\(cop\) value set to the current value of PL\_curstash\.

At run time\, each nextstate op that is executed\, sets PL\_curcop to point
to the currently executing COP\.

When a scope \(such as a sub\) is entered\, the current value of PL\_curcop
is pushed onto the context stack\. 

caller\(\) uses the saved PL\_curcop's on the context stack to extract out
the stash name\, line number etc\.

When BEGIN is called we're not actually in runtime\, so PL\_curcop points to
&PL\_compiling\, rather than the last executed cop\.

PL\_compiling gets its cop\_line etc fields updated every time the toker
reads in a new line \(or sees \#line\)\, but a 'package' declaration doesn't
update CopSTASH\(&PL\_compiling\)\, when perhaps it should??

\-\- 
In economics\, the exam questions are the same every year\.
They just change the answers\.

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Sep 12 03&#8203;:57&#8203;:32 2016\, davem wrote&#8203;:
> PL\_compiling gets its cop\_line etc fields updated every time the toker
> reads in a new line \(or sees \#line\)\, but a 'package' declaration
> doesn't
> update CopSTASH\(&PL\_compiling\)\, when perhaps it should??

PL\_curstash is used at compile time instead of CopSTASH\(&PL\_compiling\)\.  I have always assumed it was for efficiency’s sake\.  But\, come to think of it\, if PL\_compiling is a fixed buffer in the interpreter struct\, then \#defining PL\_curstash to CopSTASH\(&PL\_compiling\) would make no speed difference on threaded builds\, but I don’t know about non\-threaded\.  Then again\, I may be wrong about the reason why PL\_curstash is separate and CopSTASH\(&PL\_compiling\) is unused\.

Hmm\, maybe it’s because cops do not contain stash pointers under threads\.  Maybe the solution is to have pp\_caller check whether its cop == &PL\_compiling and use PL\_curstash instead\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @mauke

On Mon Sep 12 03&#8203;:57&#8203;:32 2016\, davem wrote&#8203;:
> On Sat\, Sep 10\, 2016 at 07&#8203;:47&#8203;:42AM \-0700\, l\.mai@&#8203;web\.de via RT wrote&#8203;:
> > On Sat Sep 10 07&#8203;:31&#8203;:47 2016\, sprout wrote&#8203;:
> > > On Sat Sep 10 02&#8203;:13&#8203;:50 2016\, mauke\- wrote&#8203;:
> > > > \#\!perl
> > > > use strict;
> > > > use warnings;
> > > >
> > > > package Mtfnpy;
> > > > \#line 100 "ABCDE"
> > > > BEGIN \{
> > > >     printf "package=%s\, file=%s\, line=%d\\n"\, caller;
> > > > \}
> > > >
> > > > \_\_END\_\_
> > > >
> > > > Output&#8203;:
> > > > package=main\, file=ABCDE\, line=102
> > > >
> > > > Expected output&#8203;:
> > > > package=Mtfnpy\, file=ABCDE\, line=102
> > >
> > > I’m not entirely sure I would expect that\.  I’m not sure what to
> > > expect\.  After all\, ‘calling’ BEGIN in this case has not started
> > > running yet\.
> >
> > I don't understand your last sentence but the reported location of
> > the call \(which is the "\}" of the BEGIN block\) is in package Mtfnpy\,
> > not main\. At minimum it's inconsistent\.
> 
> What's happening is as follows&#8203;:
> 
> At compile time\, 'package Foo' sets PL\_curstash to point to %Foo&#8203;::\.
> Then any subsequent COPs \(i\.e\. OP\_NEXTSTATE\) after that have their
> CopSTASH\(cop\) value set to the current value of PL\_curstash\.
> 
> At run time\, each nextstate op that is executed\, sets PL\_curcop to
> point
> to the currently executing COP\.
> 
> When a scope \(such as a sub\) is entered\, the current value of
> PL\_curcop
>  is pushed onto the context stack\.
> 
> caller\(\) uses the saved PL\_curcop's on the context stack to extract
> out
> the stash name\, line number etc\.
> 
> When BEGIN is called we're not actually in runtime\, so PL\_curcop
> points to
> &PL\_compiling\, rather than the last executed cop\.
> 
> PL\_compiling gets its cop\_line etc fields updated every time the toker
> reads in a new line \(or sees \#line\)\, but a 'package' declaration
> doesn't
> update CopSTASH\(&PL\_compiling\)\, when perhaps it should??

Thanks for this detailed explanation\! I discovered this issue because I have a C module that calls a Perl sub at parse time\. That Perl sub always saw the wrong calling package and I didn't understand why\. So I thought I'd check how perl itself handles that situation in BEGIN blocks \(and maybe steal the code\)\.

As it turned out\, BEGIN blocks have exactly the same issue\. Based on your explanation I've come up with the following workaround&#8203;:

&nbsp;    assert\(PL\_curcop == &PL\_compiling\);
&nbsp;    COP curcop\_with\_stash = PL\_compiling;
&nbsp;    CopSTASH\_set\(&curcop\_with\_stash\, PL\_curstash\);
&nbsp;    PL\_curcop = &curcop\_with\_stash;
&nbsp;    n = call\_sv\(sv\, G\_SCALAR\);
&nbsp;    PL\_curcop = &PL\_compiling;

It seems to pass all tests\. :\-\)

\(I don't know what the best way is to fix this in core\.\)

**Comment by p5pRT** [comment]
### From @mauke

On Sat Sep 10 02&#8203;:13&#8203;:50 2016\, mauke\- wrote&#8203;:
> \#\!perl
> use strict;
> use warnings;
> 
> package Mtfnpy;
> \#line 100 "ABCDE"
> BEGIN \{
>     printf "package=%s\, file=%s\, line=%d\\n"\, caller;
> \}
> 
> \_\_END\_\_
> 
> Output&#8203;:
> package=main\, file=ABCDE\, line=102
> 
> Expected output&#8203;:
> package=Mtfnpy\, file=ABCDE\, line=102

Here's a patch that adds a TODO test for now\.

**Comment by p5pRT** [comment]
### From @mauke

<details><summary>0001-t-op-caller.t-add-a-TODO-test-for-RT-129239.patch</summary>

```diff
From bb3329a391fce27c00dd55aeaf5357337aa09192 Mon Sep 17 00:00:00 2001
From: Lukas Mai <l.mai@web.de>
Date: Fri, 21 Oct 2016 15:18:38 +0200
Subject: [PATCH] t/op/caller.t: add a TODO test for RT #129239

---
 t/op/caller.t | 16 +++++++++++++++-
 1 file changed, 15 insertions(+), 1 deletion(-)

diff --git a/t/op/caller.t b/t/op/caller.t
index 3017465..1ffb5b3 100644
--- a/t/op/caller.t
+++ b/t/op/caller.t
@@ -5,7 +5,7 @@ BEGIN {
     chdir 't' if -d 't';
     require './test.pl';
     set_up_inc('../lib');
-    plan( tests => 97 ); # some tests are run in a BEGIN block
+    plan( tests => 100 ); # some tests are run in a BEGIN block
 }
 
 my @c;
@@ -352,3 +352,17 @@ EOP
 $::testing_caller = 1;
 
 do './op/caller.pl' or die $@;
+
+{
+    package RT129239;
+    BEGIN {
+        my ($pkg, $file, $line) = caller;
+        ::is $file, 'virtually/op/caller.t', "BEGIN block sees correct caller filename";
+        ::is $line, 12345,                   "BEGIN block sees correct caller line";
+        TODO: {
+            local $::TODO = "BEGIN blocks have wrong caller package [perl #129239]";
+            ::is $pkg, 'RT129239',               "BEGIN block sees correct caller package";
+        }
+#line 12345 "virtually/op/caller.t"
+    }
+}
-- 
2.10.0


```
</details>

**Comment by p5pRT** [comment]
### From @KES777

This issue seems is related&#8203;: https://rt.perl.org/Public/Bug/Display.html?id=127083
to the current one\.

**Comment by mauke** [comment]
This is a duplicate of https://github.com/Perl/perl5/issues/15109.
Fixed in 5.36.0.
